Muxer selection in security handshake (#446)
libp2p/specs: connections/muxer-sel-in-sec-handshake.md

found, the multiplexer negotiation process MUST fall back to multistream
-selection protocol.

Example: Noise handshake between peers that have a mutually supported
multiplexer.
    Initiator supports: ["/yamux/1.0.0", "/mplex/6.7.0"]
    Responder supports: ["/mplex/6.7.0", "/yamux/1.0.0"]

    XX:
    -> e
    <- e, ee, s, es, ["/mplex/6.7.0", "/yamux/1.0.0"] 
    -> s, se, ["/yamux/1.0.0", "/mplex/6.7.0"] 

    After handshake is done, both parties can arrive on the same conclusion
    and select "/mplex/6.7.0" as the multiplexer to use.

Example: Noise handshake between peers that don't have mutually supported
multiplexers.
    Responder supports: ["/mplex/6.7.0"]
    Initiator supports: ["yamux/1.0.0"]

    XX:
    -> e
    <- e, ee, s, es, ["/mplex/6.7.0"]
    -> s, se, ["yamux/1.0.0"]
    
    After handshaking is done, early data processing will find no mutually
    supported multiplexer, and falls back to multistream-selection protocol.

The multiplexer selection logic SHOULD run after the Noise handshake has
finished mutual authentication of the peers to enhance security.

The format of the early data is specified in [Noise-handshake-payload]

> In the previous versions that do not support this feature, the Noise handshake
> messages carry an empty field of muxer list in the early data extension. When a
> version that supports this feature talks to an older version which does not
> support this feature, the multiplexer selection process on both peers will
> generate an empty list and they MUST fall back to the multistream-selection
> process to ensure backward compatibility.

Reviewer: We already specified this before.
```suggestion
```

> ## Security

Reviewer: What you're discussing here is privacy, not security.

The multiplexer list carried in the TLS ALPN extension field is part of the
ClientHello message which is not encrypted. This feature will expose the
supported multiplexers in plain text, but this is not a weakening of security

> posture. In the future when [ECH] is ready the multiplexer info can be protected

Reviewer: It definitely is a weakening!
I don't think we should reference ECH here, as we don't have a deployment path for that protocol. Instead, point out that muxers are not really private information, since it's trivial to obtain that information by connecting to the node.

too.

> The early data in Noise handshake is only sent after the peers establish a
> shared key, in the second and third handshake messages in the XX pattern. So the
> early data is encrypted and the multiplexer info carried over is protected.
> There is no security weakening in this case either.

Reviewer: Privacy, not security!

```suggestion
Early data in the Noise handshake is sent after the peers have established a
shared key, so an on-path observer won't be able to read the early data.
```

muxer of a connection. The solution relies on the ability of the security
protocol's handshake process to negotiate higher level protocols, which enables
the muxer selection to be carried out along with security protocol handshake.
The proposed solution saves the RTT of multistream selection for muxers.

For more context and the status of this work, please refer to [#426]

## Design

### Current connection upgrade process

The current connection upgrade process is described in detail in [connections].
As shown in this [sequence-chart], after a network connection is established,
the following will happen to upgrade the conntection to a capable connection.

1. The multistream-selection protocol is run over the connection to select the
security protocol to be used.
2. The selected security protocol performs handshaking and establishs a secure
tunnel
3. The multistream-selection protocol then will run again for muxer selection.
4. Connection then is upgraded to a capable connection by the selected muxer.

## Improved muxer selection

The security protocol's ability of supporting higher level abstract protocol

> negotiation (for example, TLS's support of NextProtos, and Noise's support of

Author: yes, good point.

   be accomplished within the TLS handshake, without adding network
   round-trips, and allows the server to associate a different
   certificate with each application protocol, if desired.

For the purpose of muxer selection, the types of muxers are coded as protocol
names in the form of a list of strings, and inserted in the ALPN "NextProtos"
field. An example list as following:

    ["yamux/1.0.0", "mplux", "libp2p"]

The NextProtos list is ordered by preference, with the most prefered muxer at
the beginning. The "libp2p" protocol code MUST always be the last item in the
ALPN NextProtos list.

The server chooses the supported protocol by going through its prefered
protocol list and searchs if the protocol is supported by the client too. if no
mutually supported protcol is found the TLS handshake will fail.

If the selected NextProto is "libp2p" then the muxer selection process returns
an empty result, and the multistream-selection protocol MUST be run to negotiate
the muxer.

(TBD: in some cases early abortion is desireable)

### Muxer selection over Noise

The libp2p Noise implementation allows the Noise handshake process to carry

> early data. [Noise-Early-Data] is carried in the second and third message of

Reviewer: We need to specify which one exactly.

Author: sure I can add more specifics on handshaking/early data interaction.

Reviewer: There'd be an advantage to using the first and the second handshake message instead of the second and the third: After receiving the list of muxers from the client, the server would already know which muxer was chosen, and could already initialize that muxer and start opening streams.
This would allow the server to optimistically (at this point, the client is not yet authenticated, so we can't trust its peer ID) send some data.

For example, I could see us opening an identify stream to push our list of supported protocols at this point. We don't really care if this list is captured by a MITM, but it would allow the client to proceed faster after the handshake.

What do you think?

Reviewer 2: does the data actually get sent over the wire? I thought the write would be blocked until the noise handshake finishes.

Author: Given TLS already has muxer info in the plain text for the first TLS handshake message, and the fact muxer info is not sensitive data, I don't have objections to transmit the muxer info in the first and second Noise handsahke messages.

Reviewer 2: > the server would already know which muxer was chosen, and could already initialize that muxer and start opening streams.
This would allow the server to optimistically (at this point, the client is not yet authenticated, so we can't trust its peer ID) send some data.

I don't think we could do this if we are encrypting the connection with the key derived from the noise handshake. The data that's sent after the handshake completes must be derived from all the DH operations. If the server optimistically starts sending data (e.g. initializing the muxer and opening streams) it will be encrypted with the wrong key after the noise handshake completes. The server needs to wait for the 3rd handshake message (in XX) for it to know what key to use to encrypt the data over the wire.

Author: Per the discussion on #454, we decided to use first and second messages to send early data, so resolving this conversation thread.

the handshake pattern:

    XX:
    -> e
    <- e, ee, s, es
    -> s, se

At the end of the handshake pattern, both the client and server have received
the peer's early data. The Noise protocol does not perform the protocol
selection as TLS does, rather it just delivers the early data to handshaking
peers.

The muxer selection logic runs out of the Noise handshake process, relying on
the early data exchanged during the handshake. The early data is delivered in
the form of a byte string. The supported muxers are passed in space separated

> string codes. An example early data string:

Reviewer 3: I am very much in favor of using a protobuf here for the above mentioned reasons.

How about taking this one step further and use the _Protocol Select_ (https://github.com/libp2p/specs/pull/349) protobuf definition? To me this proposal actually reads as an extension to _Protocol Select_. Would people agree with that?

> string codes. An example early data string:

Reviewer: I’m not convinced that would be a good idea.

1. This is specific to just the Noise handshake, there’s no reason to cause any coupling here
2. The Protocol Select spec might change before we actually deploy it, which would leave us with a zombie protobuf here

> string codes. An example early data string:

Reviewer 3: > This is specific to just the Noise handshake

Related to this discussion, would it not make sense to use the same structure for both Noise and TLS? In case TLS only allows valid UTF-8 strings, we can use `multibase`.

> string codes. An example early data string:

Reviewer 3: > The Protocol Select spec might change before we actually deploy it, which would leave us with a zombie protobuf here

In case it changes, we can do those changes in a backwards compatible way, given that we are using Protobuf, no?

> string codes. An example early data string:

Reviewer 3: I am still failing to see the difference between the Protocol Select protocol and this proposal. The only difference to me is that this proposal exchanges the messages via the early data mechanism, whereas the Protocol Select protocol exchanges the messages on an _established_ channel. Am I missing something?

Both protocols
- Exchange a list of protocols
- Use the same prioritization algorithm
- Fail in the same scenarios (no protocols overlap)

How about adding an optional early-negotiation mechanism to Protocol Select, namely the exchange described in this pull request.

> string codes. An example early data string:

Reviewer: > Related to this discussion, would it not make sense to use the same structure for both Noise and TLS? In case TLS only allows valid UTF-8 strings, we can use multibase.

That doesn't work with existing libraries, who do ALPN negotiation for you. They expect a certain format.

> In case it changes, we can do those changes in a backwards compatible way, given that we are using Protobuf, no?

I still don't see the advantage of using a more complicated than necessary Protobuf here. This negotiation step is very tightly bound to the handshake protocol, and it's reasonable to expect that different handshake protocols will use different encodings (as TLS and Noise already do).

> The only difference to me is that this proposal exchanges the messages via the early data mechanism, whereas the Protocol Select protocol exchanges the messages on an established channel. Am I missing something?

That's a huge difference. It saves 1-RTT *in every case*. In specific, early data in TLS 1.3 only works on resumed connections, whereas this proposal works on every connection.

> string codes. An example early data string:

Author: I like this section of discussion, it clarifies lot of things. 
I will add a protobuf definition for Noise.

> string codes. An example early data string:

Reviewer 3: > > The only difference to me is that this proposal exchanges the messages via the early data mechanism, whereas the Protocol Select protocol exchanges the messages on an established channel. Am I missing something?
> 
> That's a huge difference. It saves 1-RTT _in every case_. In specific, early data in TLS 1.3 only works on resumed connections, whereas this proposal works on every connection.

Sorry. I think I phrased this in a confusing way. I am definitely in favor of the 1RTT saving. My point was that I don't see the difference in the two that would justify different encodings. With the restriction on TLS' ALPN mechanism, this makes sense now.

> > Related to this discussion, would it not make sense to use the same structure for both Noise and TLS? In case TLS only allows valid UTF-8 strings, we can use multibase.
> 
> That doesn't work with existing libraries, who do ALPN negotiation for you. They expect a certain format.

Thanks for raising this. I was not aware of (/ didn't think of) this restriction. It is the case for Rust, I am guessing it is also the case for the other languages and their major TLS implementations.

Let's use the current proposal and not the Protocol Select Protobuf.

    "yamux/1.0.0 mplux"

The byte string is ordered by preference, with the most prefered muxer at the
beginning.

After the Noise handshake, the client and server run the muxer selection
process with the same logic. Each side will go through the server's early
data from most prefered to lest prefered muxer, and if the muxer is in the
client's early data list, that muxer is selected. The process guarantees that
both the client and server reaches at the same conclusion of muxer slection.

If the muxer selection process does not find any mutually supported muxer, for
example, in the case that one early data string is empty, then an empty muxer
selection result is returned, and multistream-selection MUST be performed.

(TBD: in some cases early abortion is desireable)

## Cross version support

The improved muxer selection approach MUST be interoperable with pervious
libp2p versions which do not support this improved approach.

### TLS case

In the current version of libp2p, the "NextProtos" field is populated with a

> key "libp2p". By appending the key of "libp2p" to the end of the supported

Author: That's correct, I double checked the go implementation, Here is what it says: 
        // NextProtos is a list of supported application level protocols, in
	// order of preference. If both peers support ALPN, the selected
	// protocol will be one from this list, and the connection will fail
	// if there is no mutually supported protocol. If NextProtos is empty
	// or the peer doesn't support ALPN, the connection will succeed and
	// ConnectionState.NegotiatedProtocol will be empty.

The connection fails if there is not mutually supported ALPN, and can succeed if one of the ALPN fields is empty.  So we don't have to append "libp2p".

> key "libp2p". By appending the key of "libp2p" to the end of the supported

Reviewer 2: We should probably document this. I don't think we have anything about this in this repo.

> key "libp2p". By appending the key of "libp2p" to the end of the supported

Reviewer: >  If both peers support ALPN, the selected
> protocol will be one from this list, and the connection will fail
> if there is no mutually supported protocol.

I'm surprised by this. I _thought_ that the Go standard library didn't enforce this. RFC 7301 doesn't require peers to  fail the handshake if ALPN negotiation fails, which is why we had to require that behavior in QUIC (see https://datatracker.ietf.org/doc/html/rfc9001#section-8.1).

As @MarcoPolo said, I agree that it would be valuable to be very precise about what we expect in various case (ALPN offered with muxers, APLN offered with only libp2p, no / empty ALPN offered), and to have tests for this behavior in our implementation.

> key "libp2p". By appending the key of "libp2p" to the end of the supported

Author: Of course we will create test cases for this. It seems there are some differences between the RFC and the go lib implementation documentation. We need to flush out what is the actual behavior.

is always satisfied. When one peer runs the old version and the other peer runs
the version that supports this feature, the negotiated protocol is "libp2p".

In the case "libp2p" is the result of TLS ALPN, an empty result MUST be
returned to the upgrade process to indicate that no muxer was selected. And the
upgrade process MUST fall back to the multistream-selection protocol to
to negotiate the muxer to be selected.

### Noise case

The existing version of libp2p Noise handshake carries empty early data. When a
version that supports this feature talks to an older version which does not
support this feature, the muxer selection process on the new version runs
against an empty string and will return empty muxer selection result.

In the case an empty muxer selection result is returned, the upgrade process
MUST fall back to the multistream-selection protocol to select the muxer.

## Security

The muxer list carried in TLS NextProtos field is part of the ClientHello
message which is not encrypted. This feature will expose the supported muxers
in plain text, but this is not a weakening of securiy posture. In the fuure
when [ECH] is ready the muxer info can be protected too.

> The early data in Noise handshake is encrypted so that the muxer info carried

Author: For the Noise case, we plan to send early data in the second and third message in the xx pattern, (which I believe the libp2p implementation uses), those messages are encrypted so the early data can be encrypted too.

# Muxer selection in security handshake

| Lifecycle Stage | Maturity      | Status | Latest Revision |
|-----------------|---------------|--------|-----------------|
| 1A              | Working Draft | Active | r1, 2022-09-07  |

Authors: [@julian88110]

Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]

[@marten-seemann]: https://github.com/marten-seemann
[@marcopolo]: https://github.com/marcopolo
[@mxinden]: https://github.com/mxinden
[@julian88110]: https://github.com/julian88110

See the [lifecycle document][lifecycle-spec] for context about maturity level
and spec status.

[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md

## Table of Contents

- [Muxer selection in security handshake](#Muxer-selection-in-security-handshake)
    - [Table of Contents](#table-of-contents)
    - [Overview](#overview)

>     - [Design])(#design)

Author: Done, thanks!

If the client and server agree upon the common muxer to be used, then the
result of the muxer selection is a muxer code represented by the selected
protocol name. If no agreement is reached upon by the client and server
then an empty muxer code is returned and the connection upgrade process
MUST fall back to the multistream-selection protocol to negotiate the muxer.

### Muxer selection over TLS

When the security protocol selected by the upgrader is TLS, the [ALPN]
extesion of TLS handshake is used to select the muxer.

   With ALPN, the client sends the list of supported application
   protocols as part of the TLS ClientHello message.  The server chooses
   a protocol and sends the selected protocol as part of the TLS
   ServerHello message.  The application protocol negotiation can thus
   be accomplished within the TLS handshake, without adding network
   round-trips, and allows the server to associate a different
   certificate with each application protocol, if desired.

For the purpose of muxer selection, the types of muxers are coded as protocol
names in the form of a list of strings, and inserted in the ALPN "NextProtos"
field. An example list as following:

>     ["yamux/1.0.0", "mplux", "libp2p"]

Reviewer: The complete ID for yamux is `/yamux/1.0.0`: https://github.com/libp2p/go-libp2p/blob/aa6d051e651fb1fc42c7f4d98023a3b6c01df6a0/defaults.go#L32-L36

Similarly, mplex is `/mplex/6.7.0`, see for example https://github.com/libp2p/go-libp2p/blob/aa6d051e651fb1fc42c7f4d98023a3b6c01df6a0/examples/ipfs-camp-2019/08-End/main.go#L44-L47

>     ["yamux/1.0.0", "mplux", "libp2p"]

Author: Good catch, thanks to Marco and Marten.

If the selected NextProto is "libp2p" then the muxer selection process returns
an empty result, and the multistream-selection protocol MUST be run to negotiate
the muxer.

### Muxer selection over Noise

The libp2p Noise implementation allows the Noise handshake process to carry
early data. [Noise-Early-Data] is carried in the second and third message of
the handshake pattern as illustrated in the following message sequence chart.

    XX:
    -> e
    <- e, ee, s, es, _early-data_
    -> s, se, _early-data_

At the end of the handshake pattern, both the client and server have received
the peer's early data. The Noise protocol does not perform the protocol
selection as TLS does, rather it just delivers the early data to handshaking
peers.

The muxer selection logic runs out of the Noise handshake process, relying on
the early data exchanged during the handshake. The early data is delivered in
the form of a byte string in the second and third message of the XX handshake
pattern. 
 
The early data for this purpose is specified in the protobuf in the
[Early-data-specification] section. The muxers are ordered by preference, with
the most prefered muxer at the beginning.

> After the Noise handshake, the client and server run the muxer selection

Author: Great insight. Also added two examples to make the Noise handshake process more clear.

support this feature, the muxer selection process on the new version runs
against an empty string and will return empty muxer selection result.

In the case an empty muxer selection result is returned, the upgrade process
MUST fall back to the multistream-selection protocol to select the muxer.

## Security

The muxer list carried in TLS NextProtos field is part of the ClientHello
message which is not encrypted. This feature will expose the supported muxers
in plain text, but this is not a weakening of securiy posture. In the fuure
when [ECH] is ready the muxer info can be protected too.

The early data in Noise handshake is only sent afer the peers establish a
shared key, in the second and third handshake messages in the XX pattern. So
the early data is encrypted and the muxer info carried over is protected.
These is no security weakening in this case either.

## Protocol coupling

This feature aggregates the multistream-selecion function and security
handshake function. From function separation point of view, it introduces
coupling between different functions. But the argument is that in the case of
libp2p, the muxer and security are always needed at the same time, and it is a

> small price to pay to gain efficiency by ruducing one RTT.

Reviewer 2: ```suggestion
small price to pay to gain efficiency by reducing one RTT.
```

Author: done, thanks!

the security protocol's handshake process as a list of protocol names.

If the client and server agree upon the common muxer to be used, then the
result of the muxer selection is a muxer code represented by the selected
protocol name. If no agreement is reached upon by the client and server
then an empty muxer code is returned and the connection upgrade process
MUST fall back to the multistream-selection protocol to negotiate the muxer.

### Muxer selection over TLS

When the security protocol selected by the upgrader is TLS, the [ALPN]
extesion of TLS handshake is used to select the muxer.

   With ALPN, the client sends the list of supported application
   protocols as part of the TLS ClientHello message.  The server chooses
   a protocol and sends the selected protocol as part of the TLS
   ServerHello message.  The application protocol negotiation can thus
   be accomplished within the TLS handshake, without adding network
   round-trips, and allows the server to associate a different
   certificate with each application protocol, if desired.

For the purpose of muxer selection, the types of muxers are coded as protocol

> names in the form of a list of strings, and inserted in the ALPN "NextProtos"

Reviewer: “NextProtos” is the name of the field in the Go implementation (for historic reasons, because it was first an implementation of NPN, not ALPN). I suggest sticking with the RFC terminology here.

Author: good point, changed that to "ALPN extesion"

Author: thanks for the catch, agreed. changed to "ALPN extension" which is aligned with the RFC

field. An example list as following:

>     ["yamux/1.0.0", "/mplex/6.7.0", "libp2p"]

Reviewer: ```suggestion
    ["/yamux/1.0.0", "/mplex/6.7.0", "libp2p"]
```

The NextProtos list is ordered by preference, with the most prefered muxer at
the beginning. The "libp2p" protocol code MUST always be the last item in the

> ALPN NextProtos list.

Reviewer 3: ```suggestion
ALPN NextProtos list. See [#tls-case] for details on the special "libp2p" protocol code.
```

Would have helped me a bit on a first read.

an empty result, and the multistream-selection protocol MUST be run to negotiate
the muxer.

### Muxer selection over Noise

The libp2p Noise implementation allows the Noise handshake process to carry
early data. [Noise-Early-Data] is carried in the first and second message of
the XX handshake pattern as illustrated in the following message sequence chart.
The first message carries early data in the form of a list of muxers supported
by the initiator, ordered by preference. The responder sends its supported
muxer list in the second message to the initiator. After the Noise handshake
process is fully done, the initiator and responder will both process the
received eraly data and select the muxer to be used, they both iterate through
the initiator's prefered muxer list in order, and if any muxer is also
supported by the responder, that muxer is selected. If no mutually supported
muxer is found, the muxer selection process MUST fall back to multistream
-selection protocol.

Example: Noise handshake between peers that have a mutually supported muxer.
    Initiator supports: ["yamux/1.0.0", "/mplex/6.7.0"]
    Responder supports: ["yamux/1.0.0"]

    XX:

>     -> e ["yamux/1.0.0", "/mplex/6.7.0"]

Reviewer 3: Ah, I though we settled on sending these in the second and third message? Sorry in case I am confusing the past discussions. Would appreciate any pointers.

This is also how I read the section below:

> The early data in Noise handshake is only sent afer the peers establish a
shared key, in the second and third handshake messages in the XX pattern. So
the early data is encrypted and the muxer info carried over is protected.
These is no security weakening in this case either.

Author: Hi Max, thanks for the catch. Yes, you are right. I missed changing this part.

[@marcopolo]: https://github.com/marcopolo
[@mxinden]: https://github.com/mxinden
[@julian88110]: https://github.com/julian88110

See the [lifecycle document][lifecycle-spec] for context about maturity level
and spec status.

[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md

## Table of Contents

- [Muxer selection in security handshake](#Muxer-selection-in-security-handshake)
    - [Table of Contents](#table-of-contents)
    - [Overview](#overview)
    - [Design](#design)
        - [Current connection upgrade process](#current-connection-upgrade-process)
        - [Improved muxer selection](#improved-muxer-selection)
            - [Muxer selection over TLS](#muxer-selection-over-tls)
            - [Muxer selection over Noise](#muxer-selection-over-noise)
                - [Early data specification](#early-data-specification)
    - [Cross version support](#cross-version-support)
        - [TLS case](#tls-case)
        - [Noise case](#noise-case)
    - [Security](#security)
    - [Protocol coupling](#protocol-coupling)
    - [Alternative options considered](#alternative-options-considered)

> ## Overview

Author: done

This document discribes an imporvement on the connection upgrade process. The
goal of the improvement is to reduce the number of RTTs that takes to select the
muxer of a connection. The solution relies on the ability of the security
protocol's handshake process to negotiate higher level protocols, which enables
the muxer selection to be carried out along with security protocol handshake.
The proposed solution saves the RTT of multistream selection for muxers.

> For more context and the status of this work, please refer to [#426]

Author: done

## Design

### Current connection upgrade process

The current connection upgrade process is described in detail in [connections].
As shown in this [sequence-chart], after a network connection is established,
the following will happen to upgrade the conntection to a capable connection.

1. The multistream-selection protocol is run over the connection to select the
security protocol to be used.
2. The selected security protocol performs handshaking and establishs a secure
tunnel
3. The multistream-selection protocol then will run again for muxer selection.

> 4. Connection then is upgraded to a capable connection by the selected muxer.

Author: done

## Improved muxer selection

The security protocol's ability of supporting higher level abstract protocol
negotiation (for example, TLS's support of ALPN, and Noise's support of Early
Data) makes it possible to collapse the step 2 and step 3 in the previous
section into one step. Muxer selection can be performed as part of the security
protocol handshake, thus there is no need to perform another mutistream
-selection negotiation for muxer selection.

In order to achieve the above stated goal, each candidate muxer will be
represented by a protocol name/code, and the candidate muxers are supplied to
the security protocol's handshake process as a list of protocol names.

If the client and server agree upon the common muxer to be used, then the
result of the muxer selection is a muxer code represented by the selected
protocol name. If no agreement is reached upon by the client and server
then an empty muxer code is returned and the connection upgrade process

> MUST fall back to the multistream-selection protocol to negotiate the muxer.

Author: 1. That makes sense, got rid of the implementation details.
2. The behavior of falling back to multistream-selection instead of failing handshaking is a way of providing backward compatibility and also enhance security. For TLS the backword version carries non-empty ASPN, and for Noise early data is empty. And as the data is carried in variable size extension fields, I prefer to process the data only after the peers are authenticated to enhance security. Instead of introducing more edge cases and optimize for a transient phase, I prefer to simply fallback to multistream-selection in the current situation, if the negotiation is set to fail, it does not affect performace if it fails one RTT later. Open to hear other thoughts.

> ### Muxer selection over TLS

Author: done

extesion of TLS handshake is used to select the muxer.

   With ALPN, the client sends the list of supported application
   protocols as part of the TLS ClientHello message.  The server chooses
   a protocol and sends the selected protocol as part of the TLS
   ServerHello message.  The application protocol negotiation can thus
   be accomplished within the TLS handshake, without adding network
   round-trips, and allows the server to associate a different
   certificate with each application protocol, if desired.

For the purpose of muxer selection, the types of muxers are coded as protocol
names in the form of a list of strings, and inserted in the ALPN extension
field. An example list as following:

    ["/yamux/1.0.0", "/mplex/6.7.0", "libp2p"]

The muxer list is ordered by preference, with the most prefered muxer at
the beginning. The "libp2p" protocol code MUST always be the last item in the
muxer list . See [#tls-case] for details on the special "libp2p" protocol code.

The server chooses the supported protocol by going through its prefered
protocol list and searchs if the protocol is supported by the client too. if no

> mutually supported protcol is found the TLS handshake will fail.

Reviewer: Use normative language here.

Author: done

Author: done

If the selected item from the muxer list is "libp2p" then the muxer selection
process returns an empty result, and the multistream-selection protocol MUST be
run to negotiate the muxer.

### Muxer selection over Noise

> The libp2p Noise implementation allows the Noise handshake process to carry

Reviewer: Specification, not implementation.

Author: done

early data. [Noise-Early-Data] is carried in the second and third message of
the XX handshake pattern as illustrated in the following message sequence chart.
The second message carries early data in the form of a list of muxers supported

> by the responder, ordered by preference. The initiator sends its supported

Reviewer: Let's be more specific here. There are 4 messages in the server's 1st flight.

Author: Not sure I am following here. I am using the terminology found in the Noise spec https://github.com/libp2p/specs/tree/master/noise#xx, which refers to its messages by count. what is the 4th message here?

muxer list in the third message to the responder. After the Noise handshake
process is fully done, the initiator and responder will both process the
received eraly data and select the muxer to be used, they both iterate through
the initiator's prefered muxer list in order, and if any muxer is also
supported by the responder, that muxer is selected. If no mutually supported
muxer is found, the muxer selection process MUST fall back to multistream
-selection protocol.

Example: Noise handshake between peers that have a mutually supported muxer.
    Responder supports: ["yamux/1.0.0", "/mplex/6.7.0"]

>     Initiator supports: ["yamux/1.0.0"]

Reviewer: Better: use an example that has [ mplex, yamux ], so that the preference becomes clear.

Author: done

    XX:
    -> e
    <- e, ee, s, es, ["yamux/1.0.0", "/mplex/6.7.0"] 
    -> s, se, ["yamux/1.0.0"]

    After handshake is done, both parties can arrive on the same conclusion
    and select "yamux/1.0.0" as the muxer to use.

Example: Noise handshake between peers that don't have mutually supported
muxers.
    Responder supports: ["/mplex/6.7.0"]
    Initiator supports: ["yamux/1.0.0"]

    XX:
    -> e
    <- e, ee, s, es, ["/mplex/6.7.0"]
    -> s, se, ["yamux/1.0.0"]
    
    After handshaking is done, early data processing will find no mutually

>     supported muxer, and falls back to multistream-selection protocol.

Reviewer: We should fail the handshake in that case. There's no point in doing another multistream negotiation.

We can use the absence of the muxer entry in the protobuf for backwards compatibility.

Author: Please refer to the item 2 on previous comment for the reason of not prefering to fail.

> The muxer selection logic is run outside of the Noise handshake process. The

Reviewer: Not sure what that means.

Author: done

format of he early data for this purpose is specified in the protobuf in the
[Early-data-specification] section.

> #### Early data specification

Reviewer: This seems to belong in the Noise doc. We shouldn't repeat the definition of the `NoiseExtensions` here, there should only be a single source of truth.

Author: removed the proto related section

The early data message is encoded in the "protobuf2" syntax as shown in the
following. The protobuf definition is an extension to [handshake-payload]. The
existing byte array early data (the "data" field) will be replaced by a
structured NoiseExtensions protobuf message. The supported muxers and selected
muxer are populated in the "stream_muxers" field. The details of the early
data message can be find in [Noise-handshake-payload]

The muxers are ordered by preference, with the most prefered muxer at the
beginning.

```protobuf
syntax = "proto2";

message NoiseExtensions {
    repeated bytes webtransport_certhashes = 1;
    repeated string stream_muxers = 2; 
}

message NoiseHandshakePayload {
  optional bytes identity_key = 1;
  optional bytes identity_sig = 2;
  optional NoiseExtensions extensions = 4;
}
```

## Cross version support

The improved muxer selection approach MUST be interoperable with pervious

> libp2p versions which do not support this improved approach.

Reviewer: That seems like a weird place for a normative sentence. We should probably say in the introduction that this mechanism is designed to be backwards compatible.

Author: relocated.

> ### TLS case

Reviewer: Explain how backwards compatibility works in the TLS section of this document?

Author: done

In the current version of libp2p, the ALPN extension field is populated with a
key "libp2p". By appending the key of "libp2p" to the end of the supported
muxer list, the TLS handshaking process is not broken when peers run different
versions of libp2p, because the minimum overlap of the peer's supported muxer
sets is always satisfied. When one peer runs the old version and the other peer
runs the version that supports this feature, the negotiated protocol is
"libp2p".

In the case "libp2p" is the result of TLS ALPN, an empty result MUST be
returned to the upgrade process to indicate that no muxer was selected. And the
upgrade process MUST fall back to the multistream-selection protocol to
to negotiate the muxer to be selected.

> ### Noise case

Reviewer: Explain how backwards compatibility works in the Noise section of this document?

Author: done

The existing version of libp2p Noise handshake carries empty early data. When a
version that supports this feature talks to an older version which does not
support this feature, the muxer selection process on the new version runs
against an empty string and will return empty muxer selection result.

In the case an empty muxer selection result is returned, the upgrade process
MUST fall back to the multistream-selection protocol to select the muxer.

## Security

The muxer list carried in TLS ALPN extension field is part of the ClientHello
message which is not encrypted. This feature will expose the supported muxers
in plain text, but this is not a weakening of securiy posture. In the fuure
when [ECH] is ready the muxer info can be protected too.

The early data in Noise handshake is only sent afer the peers establish a
shared key, in the second and third handshake messages in the XX pattern. So
the early data is encrypted and the muxer info carried over is protected.
These is no security weakening in this case either.

> ## Protocol coupling

Reviewer: That also seems to belong in the introduction, and can probably be shortened to a single sentence.

Author: done
